src/internal/NeoEsp8266UartMethod.cpp

```diff
@@ -0,0 +1,209 @@
+
+#ifdef ARDUINO_ARCH_ESP8266
+#include "NeoEsp8266UartMethod.h"
+#include <utility>
+extern "C"
+{
+    #include <eagle_soc.h>
+    #include <ets_sys.h>
+    #include <uart.h>
+    #include <uart_register.h>
+}
+
+#define UART1 1
+#define UART1_INV_MASK (0x3f << 19)
+
+// Gets the number of bytes waiting in the TX fifo of UART1
+static inline uint8_t getUartTxFifoLength()
+{
+    return (U1S >> USTXC) & 0xff;
+}
+
+static const uint8_t* esp8266_uart1_async_buf;
+static const uint8_t* esp8266_uart1_async_buf_end;
+
+NeoEsp8266Uart::NeoEsp8266Uart(uint8_t pin, uint16_t pixelCount, size_t elementSize)
+{
+    _sizePixels = pixelCount * elementSize;
+    _pixels = (uint8_t*)malloc(_sizePixels);
+    memset(_pixels, 0x00, _sizePixels);
+}
+
+NeoEsp8266Uart::~NeoEsp8266Uart()
+{
+    free(_pixels);
+
+    // Wait until the TX fifo is empty. This way we avoid broken frames
+    // when destroying & creating a NeoPixelBus to change its length.
+    while (getUartTxFifoLength() > 0)
+    {
+        yield();
+    }
+}
+
+void NeoEsp8266Uart::InitializeUart(uint32_t uartBaud)
+{
+    // Configure the serial line with 1 start bit (0), 6 data bits and 1 stop bit (1)
+    Serial1.begin(uartBaud, SERIAL_6N1, SERIAL_TX_ONLY);
+
+    // Invert the TX voltage associated with logic level so:
+    //    - A logic level 0 will generate a Vcc signal
+    //    - A logic level 1 will generate a Gnd signal
+    CLEAR_PERI_REG_MASK(UART_CONF0(UART1), UART1_INV_MASK);
+    SET_PERI_REG_MASK(UART_CONF0(UART1), (BIT(22)));
+}
+
+void NeoEsp8266Uart::UpdateUart()
+{
+    // Since the UART can finish sending queued bytes in the FIFO in
+    // the background, instead of waiting for the FIFO to flush
+    // we annotate the start time of the frame so we can calculate
+    // when it will finish.
+    _startTime = micros();
+
+    // Then keep filling the FIFO until done
+    const uint8_t* ptr = _pixels;
+    const uint8_t* end = ptr + _sizePixels;
+    while (ptr != end)
+    {
+        ptr = FillUartFifo(ptr, end);
+    }
+}
+
+const uint8_t* ICACHE_RAM_ATTR NeoEsp8266Uart::FillUartFifo(const uint8_t* pixels, const uint8_t* end)
+{
+    // Remember: UARTs send less significant bit (LSB) first so
+    //      pushing ABCDEF byte will generate a 0FEDCBA1 signal,
+    //      including a LOW(0) start & a HIGH(1) stop bits.
+    // Also, we have configured UART to invert logic levels, so:
+    const uint8_t _uartData[4] = {
+        0b110111, // On wire: 1 000 100 0 [Neopixel reads 00]
+        0b000111, // On wire: 1 000 111 0 [Neopixel reads 01]
+        0b110100, // On wire: 1 110 100 0 [Neopixel reads 10]
+        0b000100, // On wire: 1 110 111 0 [NeoPixel reads 11]
+    };
+    uint8_t avail = (UART_TX_FIFO_SIZE - getUartTxFifoLength()) / 4;
+    if (end - pixels > avail)
+    {
+        end = pixels + avail;
+    }
+    while (pixels < end)
+    {
+        uint8_t subpix = *pixels++;
+        U1F = _uartData[(subpix >> 6) & 3];
```

Comment:
I dislike the syntax sugar that this represents.  When you glance, you think why is the same variable  (U1F) being set over and over.  Its not quickly obvious that this really is pushing data into queue.  I believe there are other assessors that are more "function" like that make this quickly obvious what is happening.